stratosphere/ncm/source/ncm_content_meta_database.cpp

```diff
@@ -0,0 +1,584 @@
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::GetRequiredSystemVersion(sf::Out<u32> out_version, ContentMetaKey key) {
+        R_TRY(this->EnsureEnabled());
+        R_UNLESS(key.type == ContentMetaType::Application || key.type == ContentMetaType::Patch, ncm::ResultInvalidContentMetaKey());
+
+        const void* value = nullptr;
+        size_t value_size = 0;
+        R_TRY(GetContentMetaValuePointer(&value, &value_size, key, this->kvs));
+
+        /* Required system version is at the same offset for the patch and application extended header.
+           We use the application header for convenience. */
+        const auto ext_header = GetValueExtendedHeader<ApplicationMetaExtendedHeader>(value);
+        out_version.SetValue(ext_header->required_system_version);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::GetPatchId(sf::Out<ProgramId> out_patch_id, ContentMetaKey key) {
+        R_TRY(this->EnsureEnabled());
+        R_UNLESS(key.type == ContentMetaType::Application, ncm::ResultInvalidContentMetaKey());
+
+        const void* value = nullptr;
+        size_t value_size = 0;
+        R_TRY(GetContentMetaValuePointer(&value, &value_size, key, this->kvs));
+        const auto ext_header = GetValueExtendedHeader<ApplicationMetaExtendedHeader>(value);
+
+        out_patch_id.SetValue(ext_header->patch_id);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::DisableForcibly() {
+        this->disabled = true;
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::LookupOrphanContent(const sf::OutArray<bool> &out_orphaned, const sf::InArray<ContentId> &content_ids) {
+        R_TRY(this->EnsureEnabled());
+        R_UNLESS(out_orphaned.GetSize() >= content_ids.GetSize(), ncm::ResultBufferInsufficient());
+
+        /* Default to orphaned for all content ids. */
+        if (out_orphaned.GetSize() > 0) {
+            std::fill_n(out_orphaned.GetPointer(), out_orphaned.GetSize(), true);
+        }
+        
+        R_UNLESS(this->kvs->GetCount() != 0, ResultSuccess());
+        
+        for (auto entry = this->kvs->begin(); entry != this->kvs->end(); entry++) {
+            const auto value = entry->GetValuePointer();
+            const auto header = GetValueHeader(value);
+
+            if (header->content_count == 0) {
+                continue;
+            }
+
+            if (content_ids.GetSize() == 0) {
+                continue;
+            }
+
+            const ContentInfo* content_infos = GetValueContentInfos(value);
+            for (size_t i = 0; i < header->content_count; i++) {
+                const ContentInfo* content_info = &content_infos[i];
+
+                /* Check if any of this entry's content infos matches one of the provided content ids.
+                   If they do, then the content id isn't orphaned. */
+                for (size_t j = 0; j < content_ids.GetSize(); j++) {
```

Comment:
Not sure what's up with all the temporaries. 

Also, up above the loop:

```cpp
auto IsContentOrphaned = [](const sf::InArray<ContentId> &list, const ncm::ContentId &id) ALWAYS_INLINE_LAMBDA {
    for (size_t i = 0; i < list.GetSize(); i++) {
        if (list[i] == id) {
            return false;
        }
    }
    return true;
};
```

And then this just becomes:

```cpp
for (size_t i = 0; i < header->content_count; i++) {
    out_orphaned[i] = IsOrphanedContent(content_ids, content_infos[i].content_id);
}
```